src/apps/basicnat/basicnat.lua

```diff
@@ -0,0 +1,172 @@
+   checksum = checksum + tcplen -- end of pseudo-header
+
+   for i = transport_base, transport_base + tcplen - 2, 2 do
+      if i ~= transport_base + 16 then -- The checksum bytes are zero
+         checksum = checksum + pkt.data[i] * 0x100 + pkt.data[i+1]
+      end
+   end
+   if tcplen % 2 == 1 then
+      checksum = checksum + pkt.data[transport_base + tcplen - 1]
+   end
+   return csum_carry_and_not(checksum)
+end
+
+
+local function fix_checksums(pkt)
+   local ipchecksum = ipv4_checksum(pkt)
+   pkt.data[ipv4_base + 10] = bit.rshift(ipchecksum, 8)
+   pkt.data[ipv4_base + 11] = bit.band(ipchecksum, 0xff)
+   local transport_proto = pkt.data[ipv4_base + 9]
+   if transport_proto == proto_tcp then
+      local transport_checksum = transport_checksum(pkt)
+      pkt.data[transport_base + 16] = bit.rshift(transport_checksum, 8)
+      pkt.data[transport_base + 17] = bit.band(transport_checksum, 0xff)
+      return true
+   elseif transport_proto == proto_udp then
+      -- ipv4 udp checksums are optional
+      pkt.data[transport_base + 6] = 0
+      pkt.data[transport_base + 7] = 0
+      return true
+   else
+      return false -- didn't attempt to change a transport-layer checksum
+   end
+end
+
+-- TODO: fix the checksum
+local function set_src_ip(pkt, ip)
+   a, b, c, d = uint32_to_bytes(ip)
+   pkt.data[ipv4_base + 12] = a
+   pkt.data[ipv4_base + 13] = b
+   pkt.data[ipv4_base + 14] = c
+   pkt.data[ipv4_base + 15] = d
+   return pkt
+end
+
+-- TODO: fix the checksum
+local function set_dst_ip(pkt, ip)
+   a, b, c, d = uint32_to_bytes(ip)
+   pkt.data[ipv4_base + 16] = a
+   pkt.data[ipv4_base + 17] = b
+   pkt.data[ipv4_base + 18] = c
+   pkt.data[ipv4_base + 19] = d
+   return pkt
+end
+
+local function ip_in_net(ip, net, mask)
+  return net == bit.band(ip, mask) % 2^32
+end
+
+-- For packets outbound from the
+-- private network, the source IP address and related fields such as IP,
+-- TCP, UDP and ICMP header checksums are translated. For inbound
+-- packets, the destination IP address and the checksums as listed above
+-- are translated.
+-- TODO: make this able to deal with a range of external IPs, not just one
+local function basic_rewrite(pkt, external_ip, internal_net, mask)
+   -- Only attempt to alter ipv4 packets. Assume an Ethernet encapsulation.
+   if pkt.data[12] ~= 8 or pkt.data[13] ~= 0 then return pkt end
+   local src_ip, dst_ip = get_src_ip(pkt), get_dst_ip(pkt)
+   if ip_in_net(src_ip, internal_net, mask) then
+      set_src_ip(pkt, external_ip)
+   end
+   if ip_in_net(dst_ip, internal_net, mask) then
+      set_dst_ip(pkt, external_ip)
+   end
+   fix_checksums(pkt)
+   return pkt
+end
+
+function BasicNAT:push ()
+   local i, o = self.input.input, self.output.output
+   local pkt = link.receive(i)
+   local external_ip = str_ip_to_uint32("10.0.0.1")
+   local internal_net, mask = str_net_to_uint32("178.0.0.0/8")
+   local natted_pkt = basic_rewrite(pkt, external_ip, internal_net, mask)
```

Comment:
The IP addresses should be parametrized. It's possible to pass these vars as arguments in BasicNAT:new.

```
function BasicNAT:new(arg)
   local o = {
      index = 1,
      packets = {},
      external_ip = arg.external_ip,
      internal_net = arg.internal_net
   }
   return setmetatable(o, {__index=BasicNAT})
end
```

And then in main.lua

```lua

local external_ip = parameters[3] or "10.0.0.1"
local internal_net = parameters[4] or "178.0.0.0/8"

config.app(c, "basicnat_app", basicnat.BasicNAT, 
   { external_ip = external_ip, internal_net = internal_net })
```